src/main/java/org/wildfly/security/authz/MappedRoleMapper.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.authz;
+
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A simple mapping role mapper.
+ *
+ * Maps each role to a set of new roles. It also supports keeping/removing of mapped/non-mapped delegate roles.
+ *
+ * @author <a href="mailto:mmazanek@redhat.com">Martin Mazanek</a>
+ */
+public class MappedRoleMapper implements RoleMapper {
+
+    private final Map<String, Set<String>> reverseRoleMap;
```

Comment:
Why is this Map named reverse?

Reply:
Because the mapping function is not applied on final array of roles, which would be just transformed to different array of roles, but it works as delegating object the transformation needs to be done in opposite direction:
User app ask whether user has role "admin" -> mapper transform it to "administrator" -> it is tested whether identity has role "administrator" in the realm.

Reply:
+1 Using this reverse map makes the implementation much easier as in reality role mappers are only mapping to a specific set of roles.